engines/ic3sa.cpp

```diff
@@ -0,0 +1,875 @@
+      if (in_projection(t, to_keep)) {
+        Term val = solver_->get_value(t);
+        m[val].insert(t);
+      }
+    }
+  }
+  return ec;
+}
+
+void IC3SA::construct_partition(const EquivalenceClasses & ec,
+                                UnorderedTermSet & out_cube) const
+{
+  // now add to the cube expressing this partition
+  for (const auto & sortelem : ec) {
+    const Sort & sort = sortelem.first;
+
+    // TODO: play around with heuristics for the representative
+    //       to add disequalities over
+    //       e.g. we're not adding all possible disequalities,
+    //       just choosing a representative from each equivalence
+    //       class and adding a disequality to encode the distinctness
+
+    //       current priority is: symbol > generic term > value
+
+    // representatives of the different classes of this sort
+    TermVec representatives;
+    Term lit;
+    for (const auto & elem : sortelem.second) {
+      const Term & val = elem.first;
+      assert(val->is_value());
+
+      const UnorderedTermSet & terms = elem.second;
+      UnorderedTermSet::const_iterator end = terms.cend();
+      UnorderedTermSet::const_iterator it = terms.cbegin();
+      Term last = *it;
+      it++;
+
+      Term repr = last;
+      bool found_repr = false;
+      bool repr_val = repr->is_value();
+
+      while (it != end) {
+        const Term & term = *(it++);
+        assert(last->get_sort() == term->get_sort());
+        lit = solver_->make_term(Equal, last, term);
+        if (!lit->is_value()) {
+          // only add if not trivially true
+          out_cube.insert(lit);
+          assert(solver_->get_value(lit) == solver_true_);
+        }
+        last = term;
+
+        // TODO: see if a DisjointSet would make this easier
+        // update representative for this class
+        if (!found_repr) {
+          if (term->is_symbolic_const()) {
+            repr = term;
+            repr_val = false;
+            found_repr = true;
+          } else if (!term->is_value() && repr_val) {
+            repr = term;
+            repr_val = false;
+          }
+        }
+      }
+
+      // save the representative for this partition
+      representatives.push_back(repr);
+    }
+
+    assert(representatives.size() == sortelem.second.size());
+
+    // add disequalities between each pair of representatives from
+    // different equivalent classes
+    for (size_t i = 0; i < representatives.size(); ++i) {
+      for (size_t j = i + 1; j < representatives.size(); ++j) {
+        const Term & ti = representatives.at(i);
+        const Term & tj = representatives.at(j);
+        // should never get the same representative term from different classes
+        assert(ti != tj);
+        if (ti->is_value() && tj->is_value()) {
+          // if there are two values, don't need to express disequality
+          continue;
+        }
+        lit = solver_->make_term(Distinct, ti, tj);
+        if (!lit->is_value()) {
```

Comment:
again, is it possible to not have this?

Reply:
This one I'm not sure. It's possible we could remove it but sometimes it was adding `true`. Even if they aren't values, it might reduce to true through simplification e.g. `x + 2 == (x + 1 + 1)` would reduce to true.